### DISCUSSION TITLE
How do I use Cue with Bazel?

### DISCUSSION THREAD

**Discussion by ah-quant** [cue-lang/cue#1269] [Q&A]
Note: this is a slightly formatted copy of a [slack discussion](https://app.slack.com/client/TLUV4Q1ST/CLT3ULF6C/thread/CLT3ULF6C-1632263210.095500) and will need more refinement.

AA Karim Yesterday at 12:26 AM
Has anyone had any success with using Bazel & Cue? Any premade rules to use?

Tony Worm  1 day ago
https://github.com/cue-lang/cue/issues/163
https://github.com/tnarg/rules_cue
I don't use the above so I cannot attest to their state

Arne  1 day ago
Looks like https://github.com/seh/rules_cue/tree/next is the hottest/most recent contender. And as @seh is quite active here, maybe he can chime in (edited) 

seh  23 hours ago
I use it every day.
The updated interface lacks coverage in the README file, but I can help steer you toward the parts you need.

seh  23 hours ago
I’ve held off on merging the “next” branch to “main” until I’ve updated the documentation to match the code.

AA Karim  19 hours ago
I’ll get started with this, thanks!

AA Karim  15 hours ago
@seh Would I need a standard cue directory layout to get started with cue_instance? I have a fairly simple set of rules at the moment and I was hoping not to have to set up cue.mod et al (edited) 

seh  15 hours ago
If you don’t have a module, you don’t need cue_instance at all.
Are you using packages?

seh  15 hours ago
You’ll probably find that cue_exported_standalone_files will meet your needs.

AA Karim  15 hours ago
thanks! I didn’t think to look at the macros :man-facepalming:

AA Karim  15 hours ago
I’m not using packages yet, no. I’m coming up with a simple POC for bringing together our dev and prod configuration using CUE and we’re using Bazel internally so it has to play nice. Packages are step 2. Any pointers for when I get there?

seh  15 hours ago
The names are tortuous, but the following are available for cue export:
cue_exported_standalone_files
cue_exported_files (requires a module)
cue_exported_instance (requires a cue_instance, which requires a module)
For cue def instead, s/exported/consolidated/ above.

seh  15 hours ago
You don’t need packages until you want to hide identifiers from outsiders or you want to refine shared definitions from more than one other consuming file.

seh  14 hours ago
Once you have packages, together with a module, you can use the cue_module rule in your cue.mod directory’s BUILD.bazel file, and then you can refer to that module from cue_instance targets elsewhere in the tree, or from cue_exported_files targets that don’t need to refer to (or refine) other instances.

seh  14 hours ago
The targets in the “leaves” build a reference graph, walking the chain of parent instances back up to the module. This helps the CUE Bazel rules collect all the files that are used by the “leaf” files—the transitive closure of both the import graph and the implicit inclusion by way of instances in dominating directories.

seh  14 hours ago
The documentation on instances describes that implicit inclusion like this:
A package is evaluated within the context of a certain directory. Within this context, only the files belonging to that package in that directory and its ancestor directories within the module are combined. We call that an instance of a package.
CUECUE
Modules, Packages, and Instances
How files are organized in CUE

AA Karim  14 hours ago
Very useful, thank you. So the cue_module would be a target that I can refer to in a cue_instance s as a dependency?

seh  14 hours ago
You might be interested in this PR on the subject as well.
#188 docs/concepts: copyedit modules document
Correct the mistaken use of the -p (--package) flag to specify the package to disambiguate an import location identifier. While we're here, correct a few misspellings and touch up the punctuation in a few places.
This follows from discussion in the "help" Slack channel.
NB: This patch replaces the original cuelang/cuelang.org#174.
[cue-lang/cuelang.org](https://github.com/cue-lang/cuelang.org) | Jul 13th | Added by GitHub

AA Karim  14 hours ago
would I ultimately need to use a cue_exported_file to create a JSON/YAML target?

seh  14 hours ago
cue_instance has an “ancestor” attribute that should nominate either the parent instance (a cue_instance) within which this instance sits or the cue_module if there are no other intervening instances that are pertinent. 

seh  14 hours ago
Yes, you’d need cue_exported_files, and its “ancestor” attribute would need to refer to either another cue_instance target or your cue_module target.

seh  14 hours ago
It’s mostly feasible to detect these targets automatically, but I didn’t think that was in the spirit of Bazel expecting or demanding that we be explicit with our dependencies.

seh  14 hours ago
The _cue_module_impl function double-checks a few things:
Is the containing directory named cue.mod?
Is there a file named module.cue in that directory?

AA Karim  14 hours ago
ok so I’ll have to build a chain of cue_instance until I reach the leaf that I want to as the src to cue_exported_file?

seh  14 hours ago
In cue_exported_files, the “srcs” attribute names the files in the “leaf instance,” even if those files aren’t in a package.
The “ancestor” attribute will name the “parent” cue_instance or cue_module. If your parent or ancestor is an instance, then we trust that we’ll be able to check that instance’s ancestor, and so on, and eventually wind up at a cue_module.

seh  14 hours ago
You can see that check in the _cue_instance_impl function, around line 318.

seh  14 hours ago
Again, though, you only need all of that once you have a module and packages.

seh  14 hours ago
I forgot to mention that there are some very small examples, including one using a module available in the test/testdata/module directory tree.

AA Karim  13 hours ago
that is helpful! I saw the from the first repo there was an examples folder that disappeared. Definitely worth having that on the top level, or something in the readme :slightly_smiling_face:

seh  13 hours ago
Understood. I wanted to turn them into executable tests that could ensure that I didn’t break things as the code evolves, but that required relocating them. Perhaps some symbolic links from an examples directory would work too.

AA Karim  13 hours ago
That would work. Another option is a section in the readme that points them to the test/testdata directory to see examples. Here’s the http_archive call, in case you needed it:
```
http_archive(
    name = "com_github_seh_rules_cue",
    url = "https://github.com/seh/rules_cue/archive/cb515350a1a691058d3ccfe2d8528c8766d69b0c.zip",
    strip_prefix = "rules_cue-cb515350a1a691058d3ccfe2d8528c8766d69b0c",
    sha256 = "fac750fb438b98628d1cf6bca7d73885db9754d9517b64203cee05a164428d76",
)
```

AA Karim  12 hours ago
Thank you! Got that PoC up and running with  modules and all. The hello world example was really helpful.
Does it work well with Gazelle?

**Comment by aakarim** [comment]
I've built a small example repo using https://github.com/seh/rules_cue/tree/next .

You can find it here:

https://github.com/aakarim/monorepo-example

**Comment by aakarim** [comment]
Any ideas on best practices for tooling @seh? Does https://github.com/seh/rules_cue/tree/next set up the Bazel toolchain such that I can just point it at a bunch of `_tool.cue` files that are in a `filegroup`?

**Reply by seh** [reply]
You _could_, but I think we'd want a new Bazel rule for using the tool layer. Such rules should be [_executable rules_](https://docs.bazel.build/versions/main/skylark/rules.html#executable-rules-and-test-rules), since the tool layer can invoke external rules that yield nondeterministic results. You probably want your tools to run unconditionally on demand. Were we to instead represent these as normal, non-executable build rules, they would only run when there was no cached output available or their input changed. Also, we'd need to tell Bazel which other tools to make available to the sandboxed execution environment for use by [CUE's _tool/exec_ package](https://pkg.go.dev/cuelang.org/go@v0.4.0/pkg/tool/exec).

**Reply by aakarim** [reply]
I think whether to run unconditionally depends on the situation. There are definitely use cases for only running when certain source files have changed - K8s manifest generation, for instance. It would be great to have an option to specify runfiles/dependencies either way.

**Reply by seh** [reply]
I have written Bazel rules that worked either way as you describe, though it helps to make them two separate rules that share most of their implementation. The reason is that an executable rule relies on an action that produces the executable file—usually a short shell program—such that the executable rule's output when used as a build rule (via _bazel build_) is that executable file, and not the content that you'd expect to be emitted from your CUE-based tool.

I don't recall there being any way for a rule implementation to detect whether it's being used by _bazel build_ versus _bazel run_.

**Reply by aakarim** [reply]
+1 on the executable rule. `cue_tool()`?
